animations/tween-animations.md

```diff
@@ -0,0 +1,726 @@
+
+You can use the AnimatedWidget helper class to separate the widget
+code from the animation code in the previous example. In the refactored
+example below, LogoApp now derives from AnimatedWidget instead of
+StatefulWidget, and the AnimatedWidget doesn't need to maintain a State
+object to hold the animation. For example:
+
+<!-- skip -->
+{% prettify dart %}
+import 'package:flutter/animation.dart';
+import 'package:flutter/material.dart';
+
+class AnimatedLogo extends AnimatedWidget {
+  AnimatedLogo({Key key, Animation<double> animation})
+      : super(key: key, listenable: animation);
+
+  @override
+  Widget build(BuildContext context) {
+    final Animation<double> animation = listenable;
+    return new Center(
+      child: new Container(
+        margin: new EdgeInsets.symmetric(vertical: 10.0),
+        height: animation.value,
+        width: animation.value,
+        child: new FlutterLogo(),
+      ),
+    );
+  }
+}
+
+class LogoApp extends StatefulWidget {
+  @override
+  _LogoAppState createState() => new _LogoAppState();
+}
+
+class _LogoAppState extends State<LogoApp> with TickerProviderStateMixin {
+  AnimationController controller;
+  Animation<double> animation;
+
+  initState() {
+    super.initState();
+    controller = new AnimationController(
+        duration: const Duration(milliseconds: 2000), vsync: this);
+    animation = new Tween(begin: 0.0, end: 300.0).animate(controller);
+    controller.forward();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new AnimatedLogo(animation: animation);
+  }
+
+  @override
+  dispose() {
+    super.dispose();
+    controller.dispose();
+  }
+}
+
+void main() {
+  runApp(new LogoApp());
+}
+{% endprettify %}
+
+LogoApp passes the Animation object to the base class and uses
+`animation.value` to set the height and width of the container, so
+it works exactly the same as before.
+
+You can find the source for this example,
+[animate2,](https://raw.githubusercontent.com/InMatrix/animation_tutorial/master/lib/animate2.dart)
+on GitHub.
+
+<a name="monitoring"></a>
+## Monitoring the progress of the animation
+
+It’s often helpful to know when an animation changes state,
+such as finishing, moving forward, or reversing.
+You can get notifications for this with `addStatusListener()`.
+The following code modifies the previous example so that
+it listens for a state change and prints an update.
+The highlighted line shows the change:
+
+<!-- skip -->
+{% prettify dart %}
+class _LogoAppState extends State<LogoApp> with SingleTickerProviderStateMixin {
+  AnimationController controller;
+  Animation<double> animation;
+
+  initState() {
+    super.initState();
+    controller = new AnimationController(
+        duration: const Duration(milliseconds: 2000), vsync: this);
+    animation = new Tween(begin: 0.0, end: 300.0).animate(controller)
+      [[highlight]]..addStatusListener((state) => print("$state"));[[/highlight]]
+    controller.forward();
+  }
+  //...
```

Comment:
Can we make `//...` stand out a bit more? It's easy to copy and paste it and believe it's a self-contained code snippet.